### DISCUSSION TITLE
Dynamic Vertex Update

### DISCUSSION THREAD

**Discussion by rainergericke** [vsg-dev/VulkanSceneGraph#604] [Q&A] [answered]
Hi Robert,

in my actual project I want to use dynamic vertex update like in the example vsgdynamicvertex. The example is easy to understand and works on Mac and Windows. I have this problem now: the dynamic vertex update has no visual effect (Mac & Windows). What works correctly:

- finding the vertex array via visitor
- setting the correct values

What can inhibit the update of the modified scene graph?

A brute force method would be to remove the scene node and create a new one based on updated data, what is a safe way to do this?

Best

**Comment by robertosfield** [comment, marked as answer]
There are 3 key parts to utilizing the built-in dynamic data update support:

1. When you create you vsg::Data you must set the data->properties.dataVariance to [vsg::DYNAMIC_DATA](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/core/Data.h#L50) or DYNAMIC_DATA_TRANSFER_AFTER_RECORD.
2. The viewer.compile() must be called afterwards to locate any dyanamic data and place it into the Viewer/RecordAndSubmitTasks vsg::TransferData containers so it can be checked on each new frame and any data that has changed can be update.
3. Call data->dirty() after you update the data so that the TransferData can detect that data needs to be transferred to the GPU.

For scene graphs that you create and then compile after then initial viewer.compile() you'll need to do the compile traversal and the update the RecordAndSumbitTasks.  There is an [vsg::updateViewer(..)](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/app/Viewer.h#L148) and function that includes these updates to the TransferTask.  

If you are curious about how the updateViewer(..) handles the additions to the TransferTask it's implemented by the [updateTasks(..) method](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/app/RecordAndSubmitTask.cpp#L201).

If you are using vsg::VertexDraw and this contains dynamic data and you are using the VulkanSceneGraph-1.0.0 release then you'll need to update to VSG master or the 1.0 branch as I've found and [fixed a bug](d12ee0ff7b6c53d438df70ce5f62a0833adad9e3) in vsg::CollectResourceRequirements visitor so it properly handles the VertexDraw classes data arrays.

**Comment by rainergericke** [comment]
Thanks, it works now. I missed to set data->properties.dataVariance!

**Comment by bluehy135** [comment]


**Reply by robertosfield** [reply]
Just posting code without context and without any explanation isn't particularly helpful.  Does the code work?  Does it not?  If so what's wrong?

**Comment by hytel** [comment]
I have a related problem.  I have a VertexDraw, and I could update the array vertices, mark it dirty, and the recompile using the updateViewer as above.  Indeed, I now do that all over my code when I need to change vertex values and this works as long as I mark it dynamic.  But, one particular vsg::VertexDraw I have will contain a vertex array that might need to change size once in a while (e.g. more or fewer triangles).  I know would have to update the VertexDraw "vertexCount" when I change this array vertex count, the other arrays are not really being set, or they only have one value (e.g. normals, color, etc).  I have tried to do this by allocating a new array, and copying the new values in.  I then try to do an assignment to the "arrays" of the vsg::VertexDraw with the new vertices like: vd->arrays[0] = vsg::BufferInfo::create(new_vertices);  When I try to compile the VertexDraw now that it has new vertex array, I get a an error like this: 

Warning: Existing deviceBufferInfo, ref_ptr<vsg::BufferInfo>(vsg::BufferInfo 0000020CC4B294D8), deviceBufferInfo->range  = 28074420, 30651948 NOT compatible

I'm assuming this is because the VertexDraw had a different size before.  How can I make Vulkan happy?

**Reply by robertosfield** [reply]
I can't speculate on why what you are doing isn't working and don't have time to go diving into topics to try and figure things on your behalf.  The best I can do is provide some high level pointers and let you figure it out.

The way I'd do it is simply make the array big enough to contain the upper size you expect and then just alter the VertexDraw::vertexCount as you require.  

With VSG master you can now assign a array to the TransferTask to have it copied on the next frame with the need for setting the data.properties.dataVariance to DYNAMIC_DATA, this can work well for adhoc updates that happen rarely.
